File: 2/aoc2.py
from collections import namedtuple


pw_metadata = namedtuple('pw_metadata', ['min', 'max', 'required_char', 'password'])


def get_metadata(line):
    policy, password = line.split(': ')
    numbers, char = policy.split(' ')
    min, max = numbers.split('-')
    return pw_metadata(int(min), int(max), char, password)


def validator_one(line):
    md = get_metadata(line)
    count = len([c for c in md.password if c == md.required_char])
    return md.min <= count <= md.max


def validator_two(line):
    md = get_metadata(line)
    p1 = md.min - 1
    p2 = md.max - 1
    return ((md.password[p1] == md.required_char and md.password[p2] != md.required_char) or
            (md.password[p1] != md.required_char and md.password[p2] == md.required_char))


valid_count_one = 0
valid_count_two = 0

for line in open('passwords.txt'):
    if validator_one(line):
        valid_count_one += 1
    if validator_two(line):
        valid_count_two += 1

print('Part 1:', valid_count_one)
print('Part 2:', valid_count_two)

File: 2/aoc2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Meta struct {
	min      int
	max      int
	char     string
	password string
}

func make_meta(line string) Meta {
	policy_and_password := strings.Split(line, ": ")
	numbers_and_char := strings.Split(policy_and_password[0], " ")
	min_and_max := strings.Split(numbers_and_char[0], "-")
	min, _ := strconv.Atoi(min_and_max[0])
	max, _ := strconv.Atoi(min_and_max[1])
	return Meta{min, max, numbers_and_char[1], policy_and_password[1]}
}

func validator_one(m Meta) bool {
	count := 0
	for _, c := range m.password {
		if string(c) == m.char {
			count++
		}
	}
	return m.min <= count && count <= m.max
}

func validator_two(m Meta) bool {
	return ((string(m.password[m.min - 1]) == m.char && string(m.password[m.max - 1]) != m.char) ||
		(string(m.password[m.min - 1]) != m.char && string(m.password[m.max - 1]) == m.char))
}

func main() {
	total1 := 0
	total2 := 0
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		m := make_meta(scanner.Text())
		if validator_one(m) {
			total1++
		}
		if validator_two(m) {
			total2++
		}
	}
	fmt.Println("Part 1:", total1)
	fmt.Println("Part 2:", total2)
}
